sensor/common/networkflow/manager/purger.go

```diff
@@ -0,0 +1,284 @@
+	case common.SensorComponentEventOfflineMode:
+		if !features.SensorCapturesIntermediateEvents.Enabled() {
+			p.purgerTicker.Stop()
+		}
+	}
+}
+
+func (p *NetworkFlowPurger) run() {
+	defer p.stopper.Flow().ReportStopped()
+	for {
+		select {
+		case <-p.stopper.Flow().StopRequested():
+			return
+		case _, chanOpen := <-p.purgerTickerC:
+			p.purgingDone.Reset()
+			// Do not execute potentially-expensive purger rules when ticker channel is closed.
+			if chanOpen {
+				p.runPurger()
+			}
+			p.purgingDone.Signal()
+		}
+	}
+}
+
+func (p *NetworkFlowPurger) runPurger() {
+	if env.EnrichmentPurgerTickerCycle.DurationSetting() == 0 {
+		// purger is disabled, but we still want to consume the messages from purgerTickerC
+		return
+	}
+	if p.manager == nil {
+		log.Warn("Programmer error: network flow purger is not bound to a network flow manager. Not purging.")
+		return
+	}
+	numPurgedActiveEp := purgeActiveEndpoints(&p.manager.activeEndpointsMutex, p.maxAge, p.manager.activeEndpoints, p.clusterEntities)
+	numPurgedActiveConn := purgeActiveConnections(&p.manager.activeConnectionsMutex, p.maxAge, p.manager.activeConnections, p.clusterEntities)
+	numPurgedHostEp, numPurgedHostConn := purgeHostConns(&p.manager.connectionsByHostMutex, p.maxAge, p.manager.connectionsByHost, p.clusterEntities)
+	log.Debugf("Purger deleted: "+
+		"%d active endpoints, %d active connections, "+
+		"%d host endpoints, %d host connections",
+		numPurgedActiveEp, numPurgedActiveConn, numPurgedHostEp, numPurgedHostConn)
+}
+
+func purgeHostConns(mutex *sync.Mutex, maxAge time.Duration, enrichmentQueue map[string]*hostConnections, store EntityStore) (numPurgedEps, numPurgedConns int) {
+	timer := prometheus.NewTimer(flowMetrics.ActiveEndpointsPurgerDuration.WithLabelValues("hostConns"))
+	defer timer.ObserveDuration()
+	numPurgedEps = 0
+	numPurgedConns = 0
+	concurrency.WithLock(mutex, func() {
+		for _, c := range enrichmentQueue {
+			concurrency.WithLock(&c.mutex, func() {
+				npe, npc := purgeHostConnsNoLock(maxAge, c, store)
+				numPurgedEps += npe
+				numPurgedConns += npc
+			})
+		}
+	})
+	return numPurgedEps, numPurgedConns
+}
+
+func purgeHostConnsNoLock(maxAge time.Duration, conns *hostConnections, store EntityStore) (numPurgedEps, numPurgedConns int) {
+	numPurgedEps = 0
+	numPurgedConns = 0
+	cutOff := timestamp.Now().Add(-maxAge)
+	for endpoint, status := range conns.endpoints {
+		// remove if the endpoint is not in the store (also not in history)
+		if len(store.LookupByEndpoint(endpoint.endpoint)) == 0 {
+			delete(conns.endpoints, endpoint)
+			numPurgedEps++
+			flowMetrics.PurgerEvents.WithLabelValues("hostEndpoint", "endpoint-gone").Inc()
+			continue
+		}
+		if maxAge > 0 {
+			// finally, remove all that didn't get any update from collector for a given time
+			if cutOff.After(status.tsAdded) {
+				flowMetrics.PurgerEvents.WithLabelValues("hostEndpoint", "max-age-reached").Inc()
+				delete(conns.endpoints, endpoint)
+				numPurgedEps++
```

Comment:
nit: This is probably OCD but I find it more readable if all the places have the same order:
In some places is:
```golang
delete(conns.endpoints, endpoint)
numPurgedEps++
flowMetrics.PurgerEvents.WithLabelValues("hostEndpoint", "endpoint-gone").Inc()
```
in others:
```golang
flowMetrics.PurgerEvents.WithLabelValues("hostEndpoint", "max-age-reached").Inc()
delete(conns.endpoints, endpoint)
numPurgedEps++
```